synapse/handlers/register.py

```diff
@@ -110,6 +112,10 @@ def __init__(self, hs: "HomeServer"):
         self._server_notices_mxid = hs.config.servernotices.server_notices_mxid
         self._server_name = hs.hostname
 
+        # The set of users that we're currently pruning devices for. Ensures
+        # that we don't have two such jobs running at once.
+        self._currently_pruning_devices_for_users: Set[str] = set()
```

Comment:
```suggestion
        # The set of users that we're currently pruning devices for. Ensures
        # that we don't have two such jobs for the same user running at once.
        self._currently_pruning_devices_for_users: Set[str] = set()
```

I wonder if we should cap the number of users we allow at once? Or have a single queue we add them to and chunk through them in a background task, so we would only process a single user at once?

Reply:
Yeah, good point. I've added a basic limiter to cap it at 5 users at a time :shrug: